src/gurobi_optimods/sharpe_ratio.py

```diff
@@ -0,0 +1,89 @@
+"""
+Sharpe Ratio
+------------
+"""
+
+import math
+
+import gurobipy as gp
+import numpy as np
+import pandas as pd
+from gurobipy import GRB
+
+from gurobi_optimods.utils import optimod
+
+
+@optimod()
+def max_sharpe_ratio(cov_matrix, mu, rf_rate=0, *, create_env):
+    """
+    Solve the problem of finding a portfolio that maximizes the
+    Sharpe ratio.
+
+    Parameters
+    ----------
+    cov_matrix : ndarray or DataFrame
+        2D positive-semidefinite variance-covariance matrix :math:`\Sigma`
+    mu : ndarray or Series
+        Expected return rates :math:`\mu`
+    rf_rate : float >= 0, optional
+        Non-negative risk-free rate of return (defaults to ``0``)
+
+    Returns
+    -------
+    portfolio: ndarray or Series
+        Portfolio that maximizes the Sharpe ratio
+    ratio: float
+        Sharpe ratio of the portfolio
+    """
+    indices = None
+
+    if isinstance(cov_matrix, pd.DataFrame):
+        indices = cov_matrix.index
+        cov_matrix = cov_matrix.to_numpy()
+    elif not isinstance(cov_matrix, np.ndarray):
+        raise TypeError(f"Unknown covariance matrix type: {type(cov_matrix)}")
+
+    if cov_matrix.ndim != 2:
+        raise ValueError(
+            f"Covariance matrix should be in 2 dimensions, not {cov_matrix.ndim}"
+        )
+
+    if isinstance(mu, pd.Series):
+        if indices is None:
+            indices = mu.index
+        elif not mu.index.equals(indices):
+            raise ValueError("Indices of cov_matrix and mu are misaligned")
+        mu = mu.to_numpy()
+    elif not isinstance(mu, np.ndarray):
+        raise TypeError(f"Unknown return rates type: {type(mu)}")
+
+    if mu.ndim != 1:
+        raise ValueError(f"Return rates should be in 1 dimension, not {mu.ndim}")
+
+    if not isinstance(rf_rate, float) and not isinstance(rf_rate, int):
+        raise TypeError(f"Unknown risk-free return rate type: {type(rf_rate)}")
+    elif rf_rate < 0:
+        raise ValueError("Risk-free return rate should be non-negative")
+    elif (mu < rf_rate).all():
+        raise ValueError(
+            f"No expected returns are greater than risk-free return rate of {rf_rate}"
+        )
+
+    portfolio, ratio = _max_sharpe_ratio_numpy(cov_matrix, mu, rf_rate, create_env)
+
+    if indices is not None:
+        portfolio = pd.Series(data=portfolio, index=indices)
+
+    return portfolio, ratio
```

Comment:
Could return a dataclass containing the two elements, instead of a tuple? I am not sure which is better...

Reply:
I was on the fence about this. I chose the tuple because I was only returning two objects. However, looking at the Mean-Variance Portfolio Mod, I like the idea of also returning the estimated return and risk, which is better done via a data class.

I created a new `SharpeRatioResult` data class and changed the return type.